pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -539,15 +543,42 @@ private void commitSegmentMetadataInternal(String realtimeTableName,
       lock.unlock();
     }
 
-    // TODO: also create the new partition groups here, instead of waiting till the {@link RealtimeSegmentValidationManager} runs
+    //  Creates new partition groups here instead of waiting till the {@link RealtimeSegmentValidationManager} runs
     //  E.g. If current state is A, B, C, and newPartitionGroupMetadataList contains B, C, D, E,
-    //  then create metadata/idealstate entries for D, E along with the committing partition's entries.
-    //  Ensure that multiple committing segments don't create multiple new segment metadata and ideal state entries for the same partitionGroup
+    //  then metadata/idealstate entries for D, E are created along with the committing partition's entries.
+
+    addNewPartitionGroups(realtimeTableName, tableConfig, instancePartitions, idealState, numReplicas, streamConfig,
+        newPartitionGroupMetadataList, numPartitionGroups, segmentAssignment, instancePartitionsMap);
 
     // Trigger the metadata event notifier
     _metadataEventNotifierFactory.create().notifyOnSegmentFlush(tableConfig);
   }
 
+  /**
+   * Method is kept synchronised so that multiple committing segments don't create multiple new segment metadata
+   * and ideal state entries for the same partitionGroup
+   */
+  private synchronized void addNewPartitionGroups(String realtimeTableName, TableConfig tableConfig,
```

Comment:
Two questions:
- You are not actually updating the idealstate here, right?
- If you were to add code to update the idealstate, how would that work with a background job that is also trying to add new partitions simultaneously?

Reply:
From my testing, this code does update the ideal state. Is there anything you see as missing.
As far as background job is concerned, I'll look into it. I believe taking a lock on some object is the only way

Reply:
- The code updates ideal state, but it is not written back to zookeeper anywhere. 
- It is possible (under some race conditions) that the background job runs in another controller. Need to be careful here. Good thing is, I have added code for idealstate update (PR #6483) so that we don't create two segments with the same seq number but different timestamp. It covers all existing code path, but since this is a new code path, you should make sure that is covered. In that case, the idealstate update will fail with an exception, and you need to handle it correctly.

Reply:
Ah ok, my bad. Yes idealstate is getting updated.

Reply:
In that case, if the background job beats the segment commit logic, then there will already be a consuming segment created for the new partition, and a call to updateInstanceStatesaForNewConsumingSegment() will result in an exception. Just be sure to handle the exception correctly. It is better to throw a specific exception, so that you can catch that particular exception here.

Reply:
Thanks a lot. Will do that!